### PR TITLE
Tests and ABI wrap for readdir

### PR DESCRIPTION
**Pull request by madhurajayaraman**
b/412715489

### CODE REVIEW

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 1]
```diff
@@ -0,0 +1,218 @@
+// Copyright 2015 The Cobalt Authors. All Rights Reserved.
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:1]: 25!

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 2]
```diff
@@ -17,6 +17,8 @@
 #include <algorithm>
 #include "starboard/common/log.h"
 
+static thread_local struct musl_dirent g_musl_results;
```
Comment (andrewsavage1) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:20]: I think this could and should actually be declared in the __abi_wrap_readdir function

**Review by jellefoks — APPROVED** [review]
Just one nit to fix.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 3]
```diff
@@ -49,3 +51,31 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+  g_musl_results.d_off = result_platform->d_off;
+#endif
+  g_musl_results.d_reclen = result_platform->d_reclen;
+  g_musl_results.d_type = result_platform->d_type;
+
+  memset(g_musl_results.d_name, 0, sizeof(g_musl_results.d_name));
+  constexpr auto minlen =
+      std::min(sizeof(g_musl_results.d_name), sizeof(result_platform->d_name));
+  memcpy(g_musl_results.d_name, result_platform->d_name, minlen);
+  if (result_platform == nullptr) {
```
Comment (andrewsavage1) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:77]: Isn't this check redundant after the other check above?

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 4]
```diff
@@ -49,3 +51,31 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
   return 0;
 }
+
+struct musl_dirent* __abi_wrap_readdir(DIR* dirp) {
+  // readdir segfaults if any of those parameters are missing.
+  SB_CHECK(dirp);
+
+  memset(&g_musl_results, 0, sizeof(g_musl_results));
+  struct dirent* result_platform = {0};  // The type from platform toolchain.
+  result_platform = readdir(dirp);
```
Comment (andrewsavage1) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:61]: Can you explain why we need to wrap this function? Are the types in the returned struct not well-defined? (that would confuse me, because we don't do any casting in this function, but I don't see anything else that would mean we have to wrap it either)

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:1]: I know this file shares almost all of its code with the readdir_r test—is there a way to actually share the code between the tests? IMO it's fine to have them in the same file (`posix_directory_get_next_test.cc`) if it's easiest to do via functions in an anon namespace.

If for some reason it would take a while we could instead file a p2 to follow up on

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc (hunk 4 above)
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:61]: Addressed in the meeting, but this doc should be able to answer any other questions/dropping it here for reference in case anyone else has the same question :) [go/cobalt-posix-across-abi](http://goto.google.com/cobalt-posix-across-abi)

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:1]: `readdir_r` is deprecated, so I think I'd like to keep them separate. I also think for readability it's likely better to keep the code separate, so it's easily to tell what's failing and why at a glance. For the minimal setup we're doing, I don't know that it's necessarily worth it to combine?
We can discuss and file a p2 to follow up if we decide it's needed, but I think it's not worth it.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc (hunk 2 above)
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:20]: You're right, moving it into the function to reduce scope.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc
Comment (andrewsavage1) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:1]: Talked offline, this lgtm

Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:1]: Conclusion : keeping as is for now.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 5]
```diff
@@ -49,3 +49,31 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
 }
+
+struct musl_dirent* __abi_wrap_readdir(DIR* dirp) {
+  // readdir segfaults if any of those parameters are missing.
+  SB_CHECK(dirp);
+
+  static thread_local struct musl_dirent g_musl_results;
+  memset(&g_musl_results, 0, sizeof(g_musl_results));
+
+  struct dirent* result_platform = {0};  // The type from platform toolchain.
```
Comment (yell0wd0g) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:60]: nit: This is a pointer, not a struct, so can't do `= {0};`, need instead ` = nullptr;`

Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:60]: Ah, good catch, thanks! Had initially written it to be not a pointer but switched it, missed this line.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+#include "starboard/configuration.h"
+#include "starboard/configuration_constants.h"
+#include "starboard/nplb/file_helpers.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+typedef std::set<std::string> StringSet;
```
Comment (yell0wd0g) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:32]: nit: `typedef` is C, C++ uses `using Bla = ...;`

Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc]: Fixed.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 7]
```diff
@@ -49,3 +49,31 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+  if (!result_platform) {
+    return nullptr;
+  }
+
+#if !SB_HAS_QUIRK(INCOMPLETE_DIRENT_STRUCTURE)
+  g_musl_results.d_ino = result_platform->d_ino;
+  g_musl_results.d_off = result_platform->d_off;
+#endif
+  g_musl_results.d_reclen = result_platform->d_reclen;
+  g_musl_results.d_type = result_platform->d_type;
```
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:71]: I think @y4vor will be able to explain this with more clarity if needed, but based on https://man7.org/linux/man-pages/man3/readdir.3.html#RETURN_VALUE the return struct can be statically allocated. We use thread_local to ensure that each thread has its own dirent struct in the event that multiple functions are calling at the same time.

[the remark this answers is not in the document]
Comment (jellefoks) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:71]: The man pages has this statement about thread safety:

```
In the current POSIX.1 specification (POSIX.1-2008), readdir() is not required to be thread-safe.  However, in modern implementations (including the glibc implementation), concurrent calls to readdir() that specify different directory streams are thread-safe.  In cases where multiple threads must read from the same directory stream,  us‐
       ing readdir() with external synchronization is still preferable to the use of the deprecated readdir_r(3) function.  It is expected that a future version of POSIX.1 will require that readdir() be thread-safe when concurrently employed on different directory streams.
```

But... Thinking about this function, even in a single thread, a program can do two calls to `opendir`, and then interleave calls to `readdir`. That wouldn't work correctly in this implementation...

**Review by jellefoks — CHANGES_REQUESTED** [review]
Changing approval...

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (jellefoks) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:71]: To address that, this is an option but it does make the wrapper more complex:

I think we have to store our necessary state in the DIR type.

That means we need to make a "MUSL_DIR", give it a `struct musl_dirent` member variable and also a member variable for the platform's "DIR".

That also means we'll have to modify `opendir` and `closedir`, and even `readdir_r` to allocate, free, and use this `MUSL_DIR` instead.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 8]
```diff
@@ -49,3 +49,31 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+
+struct musl_dirent* __abi_wrap_readdir(DIR* dirp) {
+  // readdir segfaults if any of those parameters are missing.
+  SB_CHECK(dirp);
+
+  static thread_local struct musl_dirent g_musl_results;
+  memset(&g_musl_results, 0, sizeof(g_musl_results));
+
+  struct dirent* result_platform = nullptr;  // Type from platform toolchain.
+  result_platform = readdir(dirp);
```
Comment (yell0wd0g) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:61]: nit: 
```
  struct dirent* result_platform = readdir(dirp);
```
?

Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc]: Fixed.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc]: I updated the implementation to do this, let me know if it looks ok.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 9]
```diff
@@ -49,3 +50,63 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+#endif
+  dirp->musl_dir_entry->d_reclen = result_platform->d_reclen;
+  dirp->musl_dir_entry->d_type = result_platform->d_type;
+
+  memset(dirp->musl_dir_entry->d_name, 0, sizeof(dirp->musl_dir_entry->d_name));
+  constexpr auto minlen = std::min(sizeof(dirp->musl_dir_entry->d_name),
+                                   sizeof(result_platform->d_name));
+  memcpy(dirp->musl_dir_entry->d_name, result_platform->d_name, minlen);
+
+  return dirp->musl_dir_entry;
```
Comment (andrewsavage1) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:111]: To confirm, based on the spec it's okay for the return value to share the memory of the parameter passed in?

Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:111]: If you look at  Jelle's comment in https://github.com/youtube/cobalt/pull/5632#discussion_r2080474917 , this is an implementation of that to see if that suggestion works.

https://man7.org/linux/man-pages/man3/readdir.3.html#RETURN_VALUE the spec doesn't seem to disallow this - but if there's something I'm missing there let me know.

Comment (andrewsavage1) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:111]: Trying to understand the situation he's describing in which the thread_local version of readdir we had failed—does one of the test cases test that situation? If not, we should have one

Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:111]: Added a test for it in the most recent commit (commit message is "Add test for 2 files being read") - the test fails on the previous implementation of readdir and passes on this one.

**Review by jellefoks — CHANGES_REQUESTED** [review]
The approach looks good, comments are for edge cases and a placebo test.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+
+  DIR* directory = opendir(path.c_str());
+  EXPECT_TRUE(directory);
+  struct dirent* dirent;
+  dirent = readdir(directory);
+  EXPECT_FALSE(!dirent);
+  EXPECT_TRUE(closedir(directory) == 0);
+}
+
+TEST(PosixDirectoryReaddirTest, FailureOnInsufficientSize) {
```
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:207]: This was copied from the `readdir_r` tests, which were copied from the SbDirectoryGetNext tests -- this maybe all ties into Holden's discussion of creating standards for our Posix tests to make sure we're testing the necessary pieces. I'll get rid of the irrelevant bits here, but this is definitely a piece of a larger conversation -- updating the potentially outdated NPLB tests.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+TEST(PosixDirectoryReaddirTest, SunnyDayTwoFiles) {
+  std::string path1 = GetTempDir();
+  EXPECT_FALSE(path1.empty());
+  FileExists(path1.c_str());
+  DIR* directory1 = opendir(path1.c_str());
+  EXPECT_TRUE(directory1 != NULL);
+
+  std::string path2 = GetTempDir();
```
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:102]: It's not intentional - I wrongly assumed that GetTempDir would do any kind of random name generation, my bad.

The purpose of this test was for the most part to ensure that the `dirent*` returned by `readdir` are different when we call it for different directories - the previous implementation didn't assure that, and was, as a result, not reentrant. It still does that - I can compare pointer addresses instead of examining any value in the struct to make that more clear, and I'll change the test name. I suppose that also means I can just generate one directory and open it twice to create two separate musl_dirs.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+  EXPECT_FALSE(path2.empty());
+  FileExists(path2.c_str());
+  DIR* directory2 = opendir(path2.c_str());
+  EXPECT_TRUE(directory2 != NULL);
+
+  struct dirent* dirent1 = readdir(directory1);
+  struct dirent* dirent2 = readdir(directory2);
+
+  EXPECT_NE(dirent1->d_name, dirent2->d_name);
```
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:111]: Addressed in another comment.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+        directory_queue.push(entry_path);
+      }
+    }
+
+    EXPECT_TRUE(closedir(directory) == 0);
+  }
+
+  // Make sure we found all of test data directories and files.
+  EXPECT_EQ(0, paths_to_find.size());
+  for (auto it = paths_to_find.begin(); it != paths_to_find.end(); ++it) {
```
Comment (jellefoks) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:186]: This can instead be a cleaner `for (const auto& path : paths_to_find)`, right?

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 14]
```diff
@@ -49,3 +50,73 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+  }
+  musl_directory->dir = directory;
+
+  musl_dirent* musl_dir_entry = (musl_dirent*)calloc(1, sizeof(musl_dirent));
+  if (!musl_dir_entry) {
+    errno = ENOMEM;
+    free(musl_directory);
+    return nullptr;
+  }
+  musl_directory->musl_dir_entry = musl_dir_entry;
```
Comment (jellefoks) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:78]: We should also set d_reclen to the size of the allocation here. See the other comment about the artificial limit on the length of d_name.

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc [hunk 15]
```diff
@@ -49,3 +50,73 @@ int __abi_wrap_readdir_r(DIR* dirp,
...
+  memset(dirp->musl_dir_entry, 0, sizeof(musl_dirent));
+#if !SB_HAS_QUIRK(INCOMPLETE_DIRENT_STRUCTURE)
+  dirp->musl_dir_entry->d_ino = result_platform->d_ino;
+  dirp->musl_dir_entry->d_off = result_platform->d_off;
+#endif
+  dirp->musl_dir_entry->d_reclen = result_platform->d_reclen;
+  dirp->musl_dir_entry->d_type = result_platform->d_type;
+
+  memset(dirp->musl_dir_entry->d_name, 0, sizeof(dirp->musl_dir_entry->d_name));
+  constexpr auto minlen = std::min(sizeof(dirp->musl_dir_entry->d_name),
```
Comment (jellefoks) [inline starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:117]: I know this is actually pre-existing code, but this approach introduces a size limit that readdir and readdir_r themselves don't have.

In the standard, this struct doesn't have a fixed size, and the member is`char d_name[]`. As a result, the man page says that the standard notes that using "sizeof(d_name) is incorrect". This statement contains a "sizeof(d_name)". What it means is that the standard API wants to not impose a maximum name length. The actual size of the memory pointed to by the 'struct dirent' pointer is large enough to hold the length of the name instead of the "sizeof(struct dirent)".

That means that in order to wrap it without introducing such a limit, we should use (and update) `d_reclen` internally and do a `realloc()` on the dirent if it's not large enough to hold the `d_name` that is received from the system.

It's reasonable to not do that as part of this PR, but a cleanup bug should be filed to make that change and remove this artificial filename length limit.

**Review by jellefoks — CHANGES_REQUESTED** [review]

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+using StringSet = std::set<std::string>;
+
+bool FileExists(const char* path) {
+  struct stat info;
+  return stat(path, &info) == 0;
+}
+
+TEST(PosixDirectoryReaddirTest, SunnyDay) {
+  const int kNumFiles = 65;
+  ScopedRandomFile files[kNumFiles];
```
Comment (jellefoks) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:41]: Does ScopedRandomFile let you instead use the C++ style `std::array<ScopedRandomFile, kNumFiles> files` here?

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc (hunk 14 above)
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.cc:78]: b/419335613

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc (hunk 16 above)
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc]: Rewrote all of the tests, closing this comment as it's not relevant anymore.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc (hunk 13 above)
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc]: Rewrote all of the tests, closing this comment as it's not relevant anymore.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+    }
+  }
+  closedir(dir);
+  rmdir(path.c_str());
+}
+
+/*
+Scenarios that aren't tested:
+1. Invalid directory pointer (ie, readdir(invalid_pointer)) as this behavior
+crashes with a seg fault.
```
Comment (yell0wd0g) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:66]: nit: GTest should be able to verify this using
```
EXPECT_EXIT(/*code*/, testing:: KilledBySignal(SIGSEGV), "");
```

This can be added late ron, or perhaps telling the Gem to
support these kinds of test?

Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:66]: Ooh I didn't know we could do that! I'll make a note of it and add later. There are probably other functions that could benefit from this too.

Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:66]: I'll follow up on the bug later today once this is in !

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+#include "starboard/common/string.h"
+#include "starboard/nplb/file_helpers.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+// Helper function to create a temporary directory for testing
+std::string CreateTemporaryDirectory(const std::string& base_name) {
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:31]: You should be using kSbSystemPathTempDirectory for this purpose instead of a hardcoded path as is done below

**Review by andrewsavage1 — CHANGES_REQUESTED** [review]

File: starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.h [hunk 19]
```diff
@@ -33,10 +33,21 @@ typedef struct musl_dirent {
   char d_name[256];
 } musl_dirent;
 
-SB_EXPORT int __abi_wrap_readdir_r(DIR* dirp,
+typedef struct musl_dir {
```
[the remark this answers is not in the document]
Comment (jellefoks) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.h:36]: We don't use musl's implementation of any of the functions that use or return `DIR` (`opendir`/`readdir`/`closedir`), and `DIR` is only used as an opaque pointer in the interfaces of those functions.

As a result, only the `opendir`/`readdir`/`closedir` wrappers that take or return `DIR` need to know the type's implementation.

Note: musl and glibc both define DIR as `typedef struct __dirstream DIR;` and leave `struct __dirstream` as a forward declaration, blocking any knowledge about the contents to the user of the APIs. There is no need to change that.

[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/shared/modular/starboard_layer_posix_directory_abi_wrappers.h:36]: Discussed offline. Current impl is fine.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc (hunk 17 above)
Comment (yell0wd0g) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:66]: A TODO with a bug associated (unassigned is fine) should be enough.
Leaving this comment unresolved not to forget about that.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc (hunk 18 above)
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc]: Updated

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+  void SetUp() override {
+    test_dir_ = CreateTemporaryDirectory("readdir_test_dir");
+    ASSERT_EQ(mkdir(test_dir_.c_str(), 0777), 0)
+        << "Failed to create test directory: " << test_dir_;
+  }
+
+  void TearDown() override { RemoveDirectoryRecursively(test_dir_); }
+
+  std::string test_dir_;
+  int flags = O_CREAT | O_WRONLY | O_TRUNC;
```
[the remark this answers is not in the document]
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc]: Fized.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc (hunk 17 above)
Comment (madhurajayaraman) [reply starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:58]: Added on ln 59.

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+#include "starboard/common/string.h"
+#include "starboard/configuration_constants.h"
+#include "starboard/nplb/file_helpers.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+// Helper function to clean up a directory and its contents
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:31]: nit: IMO the two comments in this helper function don't add anything

File: starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+
+  // Create a subdirectory
+  std::string subdir_path = test_dir_ + "/test_subdir";
+  ASSERT_EQ(mkdir(subdir_path.c_str(), 0777), 0)
+      << "Failed to create test subdirectory.";
+
+  DIR* dir = opendir(test_dir_.c_str());
+  ASSERT_NE(dir, nullptr) << "Failed to open directory.";
+
+  std::vector<std::string> expected_entries = {".", "..", "test_file.txt",
```
Comment (andrewsavage1) [inline starboard/nplb/posix_compliance/posix_directory_get_next_readdir_test.cc:121]: If you instead use `std::unordered_set`s you will not have to sort them later

**Review by jellefoks — APPROVED** [review]
